### DISCUSSION TITLE
Pyinstaller and Pytorch

### DISCUSSION THREAD

**Discussion by zainabalthafeeri1** [pyinstaller/pyinstaller#7621] [Help] [answered]
I have a project that uses Pytorch, and I have installed Pytorch from the [official website](https://pytorch.org/get-started/locally/); when I try to run the exe generated by Pyinstaller, this error always appears. I don't know why; I tried to add torch as a hidden import in the .spec file, but this did not solve the issue.

```
None of PyTorch, TensorFlow >= 2.0, or Flax have been found. Models won't be available and only tokenizers, configuration and file/data utilities can be used.
Traceback (most recent call last):
  File "main.py", line 5, in <module>
    print(chroma_client.list_collections())
  File "chromadb/api/local.py", line 165, in list_collections
  File "chromadb/api/models/Collection.py", line 54, in __init__
  File "chromadb/utils/embedding_functions.py", line 19, in __init__
  File "sentence_transformers/SentenceTransformer.py", line 95, in __init__
  File "sentence_transformers/SentenceTransformer.py", line 840, in _load_sbert_model
  File "sentence_transformers/models/Transformer.py", line 137, in load
  File "sentence_transformers/models/Transformer.py", line 29, in __init__
  File "sentence_transformers/models/Transformer.py", line 49, in _load_model
  File "transformers/utils/import_utils.py", line 1112, in __getattribute__
  File "transformers/utils/import_utils.py", line 1100, in requires_backends
ImportError: 
AutoModel requires the PyTorch library but it was not found in your environment. Checkout the instructions on the
installation page: https://pytorch.org/get-started/locally/ and follow the ones that match your environment.
Please note that you may need to restart your runtime after installation.
```

**Comment by rokm** [comment, marked as answer]
If you check the source file where the error originates from, you will see that `transformers` requires the metadata for the `torch` package to be collected: 
https://github.com/huggingface/transformers/blob/b92abfa6e0b7053c8bede0647b69f0f1690e9cd0/src/transformers/utils/import_utils.py#L1090-L1091
and 
https://github.com/huggingface/transformers/blob/b92abfa6e0b7053c8bede0647b69f0f1690e9cd0/src/transformers/utils/import_utils.py#L48-L58

So you need to instruct PyInstaller to collect the metadata, for example by adding [`--copy-metadata torch`](https://pyinstaller.org/en/stable/usage.html#cmdoption-copy-metadata) to your PyInstaller command.

**Reply by zainabalthafeeri1** [reply]
Many thanks; it's worked; I'm wondering how I can know the packages I need to copy their metadata; is there a straightforward way? Or do I need to check each package independently?

**Reply by rokm** [reply]
Typically, the missing metadata problems are investigated on per-case basis; that is, when you get an error and see that the code that raised an error is trying to use metadata. 

We have automatic detection for use of standard `importlib_metadata` / `importlib.metadata` functions that trigger automatic collection of metadata for the package, but it didn't seem to work in this case...

**Reply by bwoodsend** [reply]
The easiest way is the empirical way. i.e. run your program and see if any _dependency not satisfied_ or _distribution not found_ errors crop up.
